// homebridge-tplink-smarthome/lib/characteristics/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const default_characteristic_1 = __importDefault(require("./default-characteristic"));
const amperes_1 = __importDefault(require("./amperes"));
const kilowatt_hours_1 = __importDefault(require("./kilowatt-hours"));
const kilowatt_volt_ampere_hour_1 = __importDefault(require("./kilowatt-volt-ampere-hour"));
const volt_amperes_1 = __importDefault(require("./volt-amperes"));
const volts_1 = __importDefault(require("./volts"));
const watts_1 = __importDefault(require("./watts"));
function characteristic(Characteristic) {
    const DefaultCharacteristic = default_characteristic_1.default(Characteristic);
    return {
        Amperes: amperes_1.default(DefaultCharacteristic),
        KilowattHours: kilowatt_hours_1.default(DefaultCharacteristic),
        KilowattVoltAmpereHour: kilowatt_volt_ampere_hour_1.default(DefaultCharacteristic),
        VoltAmperes: volt_amperes_1.default(DefaultCharacteristic),
        Volts: volts_1.default(DefaultCharacteristic),
        Watts: watts_1.default(DefaultCharacteristic),
    };
}
exports.default = characteristic;
